### PR TITLE
[9.4.x] JDG-4767 ISPN-12598 Hot Rod client cluster switch happens too soon

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.redhat.com/browse/JDG-4767

### CODE REVIEW

**Review by wburns — COMMENTED** [review]
I know this is already like this on main, but I just had a few questions.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/tcp/RoundRobinBalancingStrategy.java [hunk 1]
```diff
@@ -66,11 +66,7 @@ public SocketAddress dryRunNextServer() {
    }
 
    private SocketAddress getServerByIndex(int pos) {
-      SocketAddress server = servers[pos];
-      if (trace) {
```
[the remark this answers is not in the document]
Comment (danberindei) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/tcp/RoundRobinBalancingStrategy.java:70]: It was too noisy, `nextServer()` also logs the selected server.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelFactory.java [hunk 2]
```diff
@@ -270,6 +270,11 @@ private SocketAddress getNextServer(Set<SocketAddress> failedServers, byte[] cac
       SocketAddress server;
       lock.writeLock().lock();
       try {
+         if (failedServers != null && failedServers.containsAll(getServers(cacheName))) {
+            log.debug("All the servers are marked as failed. Cluster might have completely shut down, try reverting to the initial server list.");
```
[the remark this answers is not in the document]
Comment (danberindei) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelFactory.java:274]: Yeah, that makes sense

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/RetryOnFailureOperation.java [hunk 3]
```diff
@@ -195,14 +195,14 @@ protected Throwable handleException(Throwable cause, Channel channel, SocketAddr
    }
 
    protected void logAndRetryOrFail(Throwable e, boolean canSwitchCluster) {
-      if (retryCount < channelFactory.getMaxRetries() && channelFactory.getMaxRetries() >= 0) {
```
[the remark this answers is not in the document]
Comment (danberindei) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/RetryOnFailureOperation.java:198]: That and `ConfigurationBuilder.validate()` throws an exception if `maxRetries<0`

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelFactory.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (danberindei) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelFactory.java]: Changed to warning and i18n. Main PR is https://github.com/infinispan/infinispan/pull/9475

**Comment by wburns** [comment]
LGTM just make sure is stays consistent with whatever changes on https://github.com/infinispan/infinispan/pull/9475

**Comment by danberindei** [comment]
Superseded by #9529

**Pull request closed without merging**